### PR TITLE
introduction of undistributed dimension capability in mediator and caps

### PR DESCRIPTION
**Pull request by mvertens**
undistributed dimension capability in mediator and caps

This PR accomplishes several things:
- introduces the ability to have one undistributed dimension in the caps and the mediator
   - sizeable reduction of fd.yaml that was enabled by adding undistributed dimension
   - new names in esmFldsExhange.F90 that encapsulates undistributed dimension
- changes to all the data and data components to utilize new undistributed dimensions (e.g. for wet and dry dust fluxes)
- changes to the history and restart in the mediator to write out a separate field for each undistributed dimension
- additions for ww3 and pop caps to work correctly 
- a total rework of med_map_mod.F90 for the renormalization mapping 

Hashes:

```
cice: 7274778
clm: 420af4a
MOM6: 414c539
mosart: 9b8bb38
```

Note: cam, pop and ww3 are still in svn and are entered as tags in the Externals.cfg

Test suite: testlist_drv.xml
Test baseline: /glade/scratch/jedwards/BASELINES/mar08
    - new baselines created in /glade/scratch/jedwards/BASELINES/apr02
Test namelist changes: no
Test status: bit for bit (except for X compsets)
    Note - bfb for MOM compsets were done after using the latest MOM6 hash with the older checkout 
Fixes:None
User interface changes?: No
Update gh-pages html (Y/N)?:N
Code review:

### CODE REVIEW

File: config/cesm/config_grids_common.xml [hunk 1]
```diff
@@ -1,6 +1,16 @@
 <?xml version="1.0"?>
 
   <gridmaps>
+
+    <!-- TODO: this is commented out until this grid can be run in DEBUG model with online remapping -->
```
Comment (jedwards4b) [inline config/cesm/config_grids_common.xml]: Remove this comment - there is nothing commented out here.

File: config/cesm/config_grids_common.xml [hunk 2]
```diff
@@ -460,6 +470,15 @@
       <map name="GLC2OCN_ICE_RMAPNAME">cpl/gridmaps/gland4km/map_gland4km_to_gx3v7_nnsm_e1000r500_180502.nc</map>
     </gridmap>
 
+
```
Comment (jedwards4b) [inline config/cesm/config_grids_common.xml]: remove whitespace

File: config/cesm/config_grids_mct.xml [hunk 3]
```diff
@@ -75,13 +75,14 @@
...
       <map name="OCN2ATM_SMAPNAME">cpl/gridmaps/gx1v6/map_gx1v6_TO_fv1.9x2.5_aave.130322.nc</map>
     </gridmap>
-    <gridmap atm_grid="1.9x2.5" ocn_grid="gx1v7">
-      <map name="ATM2OCN_FMAPNAME">cpl/gridmaps/fv1.9x2.5/map_fv1.9x2.5_TO_gx1v7_aave.181205.nc</map>
-      <map name="ATM2OCN_SMAPNAME">cpl/gridmaps/fv1.9x2.5/map_fv1.9x2.5_TO_gx1v7_blin.181205.nc</map>
-      <map name="ATM2OCN_VMAPNAME">cpl/gridmaps/fv1.9x2.5/map_fv1.9x2.5_TO_gx1v7_patc.181205.nc</map>
-      <map name="OCN2ATM_FMAPNAME">cpl/gridmaps/gx1v7/map_gx1v7_TO_fv1.9x2.5_aave.181205.nc</map>
-      <map name="OCN2ATM_SMAPNAME">cpl/gridmaps/gx1v7/map_gx1v7_TO_fv1.9x2.5_aave.181205.nc</map>
-    </gridmap>
+    <!-- TODO: this is commented out until this grid can be run in DEBUG model with online remapping -->
```
Comment (jedwards4b) [inline config/cesm/config_grids_mct.xml]: why is this commented out in an mct specific file?

File: src/components/data_comps/datm/nuopc/atm_comp_nuopc.F90 [hunk 4]
```diff
@@ -425,21 +414,20 @@ subroutine InitializeRealize(gcomp, importState, exportState, clock, rc)
...
+    call shr_nuopc_methods_State_SetScalar(dble(nxg),flds_scalar_index_nx, exportState,  &
          flds_scalar_name, flds_scalar_num, rc)
     if (shr_nuopc_methods_ChkErr(rc,__LINE__,u_FILE_u)) return
 
-    call shr_nuopc_methods_State_SetScalar(dble(SDATM%nyg),flds_scalar_index_ny, exportState, &
+    call shr_nuopc_methods_State_SetScalar(dble(nyg),flds_scalar_index_ny, exportState, &
          flds_scalar_name, flds_scalar_num, rc)
     if (shr_nuopc_methods_ChkErr(rc,__LINE__,u_FILE_u)) return
-
+
```
Comment (jedwards4b) [inline src/components/data_comps/datm/nuopc/atm_comp_nuopc.F90:430]: remove extra whitespace

File: src/components/data_comps/datm/nuopc/datm_comp_mod.F90 [hunk 5]
```diff
@@ -1,6 +1,7 @@
 #ifdef AIX
```
Comment (jedwards4b) [inline src/components/data_comps/datm/nuopc/datm_comp_mod.F90:1]: We should remove these ifdef AIX lines - it's untested and stale code.

File: src/components/data_comps/datm/nuopc/datm_comp_mod.F90 [hunk 6]
```diff
@@ -1426,4 +1422,227 @@ subroutine datm_comp_run(x2a, a2x, &
...
+
+    ! input/output variables
+    type(ESMF_State)     :: importState
+    integer, intent(out) :: rc
+
+    ! local variables
+    integer :: k
+    !----------------------------------------------------------------
+
+    k = mct_aVect_indexRA(x2a, 'Sx_avsdr')
```
Comment (jedwards4b) [inline src/components/data_comps/datm/nuopc/datm_comp_mod.F90:1437]: If we put the strings into an array 
'Sx_avsdr', 'Sx_avsdf', ...
couldn't this be a simple 3 line loop replacing lines 1437 through 1507?

File: src/drivers/nuopc/cime_config/buildnml [hunk 7]
```diff
@@ -337,11 +338,14 @@ def _create_runseq(case, coupling_times):
             # for Q (aquaplanet) compsets
             runseq_input = os.path.join(input_dir, 'nuopc_runseq_Q')
 
-        elif ( (comp_atm == 'datm' and comp_ocn == "mom"  and comp_ice == "dice") or 
-               (comp_atm == 'datm' and comp_ocn == "mom"  and comp_ice == "cice") or
+        elif ( (comp_atm == 'datm' and (comp_ocn == "mom" or comp_ocn == 'pop')  and comp_ice == "dice") or
```
Comment (jedwards4b) [inline src/drivers/nuopc/cime_config/buildnml:341]: We need to put this in xml and remove it from python source.

File: src/drivers/nuopc/cime_config/config_component_cesm.xml [hunk 8]
```diff
@@ -201,7 +201,9 @@
       <value compset="_CAM\d+%WCXI">288</value>
       <value compset="_DATM%COPYALL_NPS">72</value>
       <value compset="_DATM.*_CLM">48</value>
-      <value compset="_DATM.*_DICE.*_POP2">4</value>
+      <!-- NOTE: if the following is changed from 24 -> 4 - then a new run sequence is needed since the -->
+      <!-- ocn coupling interval for this case is 24 - but the atm is longer -->
+      <value compset="_DATM.*_DICE.*_POP2">24</value>
```
Comment (jedwards4b) [inline src/drivers/nuopc/cime_config/config_component_cesm.xml:206]: this is a change that would affect the mct comp_interface as well is it not?

File: src/drivers/nuopc/cime_driver/esmApp.F90 [hunk 9]
```diff
@@ -129,7 +147,7 @@ program esmApp
        call ESMF_Finalize(endflag=ESMF_END_ABORT)
 
   call ESMF_LogWrite("ESMF_GridCompDestroy called", ESMF_LOGMSG_INFO, rc=rc)
-!  call ESMF_LogSet(flush=.true., trace=.true., rc=rc)
+  ! call ESMF_LogSet(flush=.true., trace=.true., rc=rc)
```
Comment (jedwards4b) [inline src/drivers/nuopc/cime_driver/esmApp.F90:150]: why is this line commented out?

File: src/drivers/nuopc/cime_flds/esmFldsExchange.F90 [hunk 10]
```diff
@@ -146,160 +144,208 @@ subroutine esmFldsExchange(gcomp, phase, rc)
...
+    call ESMF_LogWrite('flds_i2o_per_cat = '// trim(cvalue), ESMF_LOGMSG_INFO)
 
     !----------------------------------------------------------
     ! Initialize mapping file names
     !----------------------------------------------------------
 
     ! to atm
 
-    call NUOPC_CompAttributeGet(gcomp, name='ice2atm_fmapname', value=ice2atm_fmap, rc=rc)
+    call NUOPC_CompAttributeGet(gcomp, name='ice2atm_fmapname', value=ice2atm_fmap, isPresent=isPresent, rc=rc)
```
Comment (jedwards4b) [inline src/drivers/nuopc/cime_flds/esmFldsExchange.F90:158]: this could be in a loop

File: src/drivers/nuopc/cime_flds/esmFldsExchange.F90 [hunk 11]
```diff
@@ -348,10 +393,13 @@ subroutine esmFldsExchange(gcomp, phase, rc)
           call addfld(fldListFr(compatm)%flds, 'Sa_pbot')
           call addmap(fldListFr(compatm)%flds, 'Sa_pbot', compocn, mapbilnr, 'one', atm2ocn_smap)
 
-          do n = 1,size(iso)
-             call addfld(fldListFr(compatm)%flds, 'Sa_shum'//iso(n))
-             call addmap(fldListFr(compatm)%flds, 'Sa_shum'//iso(n), compocn, mapbilnr, 'one', atm2ocn_smap)
-          end do
+          call addfld(fldListFr(compatm)%flds, 'Sa_shum')
```
Comment (jedwards4b) [inline src/drivers/nuopc/cime_flds/esmFldsExchange.F90:396]: why did you remove the loop here?

File: src/drivers/nuopc/cime_flds/esmFldsExchange.F90 [hunk 12]
```diff
@@ -427,34 +475,27 @@ subroutine esmFldsExchange(gcomp, phase, rc)
...
+    !   - hydrophobic organic carbon dry deposition flux
+    !   - hydrophylic organic carbon wet deposition flux
+    ! to lnd: dust wet deposition flux (sizes 1-4) from atm
+    ! to lnd: dust dry deposition flux (sizes 1-4) from atm
     ! to lnd: nitrogen deposition fields from atm
     ! ---------------------------------------------------------------------
 
-    ! TODO (mvertens, 2019-12-13): the nitrogen deposition fluxes here
+    ! TODO (mvertens, 2018-12-13): the nitrogen deposition fluxes here
     ! are not treated the same was as in cesm2.0 release
```
Comment (jedwards4b) [inline src/drivers/nuopc/cime_flds/esmFldsExchange.F90:492]: comment should read: the same way as

File: src/drivers/nuopc/cime_flds/esmFldsExchange.F90 [hunk 13]
```diff
@@ -1109,62 +1112,79 @@ subroutine esmFldsExchange(gcomp, phase, rc)
...
-
-          call addfld(fldListFr(compatm)%flds, 'Faxa_snowc'//iso(n))
-          call addfld(fldListFr(compatm)%flds, 'Faxa_snowl'//iso(n))
-          call addfld(fldListFr(compatm)%flds, 'Faxa_snow' //iso(n))
-          call addfld(fldListTo(compocn)%flds, 'Faxa_snow' //iso(n))
-       end do
+       call addfld(fldListFr(compatm)%flds, 'Faxa_rainc')
+       call addfld(fldListFr(compatm)%flds, 'Faxa_rainl')
+       call addfld(fldListFr(compatm)%flds, 'Faxa_rain' )
+       call addfld(fldListTo(compocn)%flds, 'Faxa_rain' )
```
Comment (jedwards4b) [inline src/drivers/nuopc/cime_flds/esmFldsExchange.F90:1162]: mixing of fldListTo and fldListFr seems confusing and potentially error prone.  
Can we have two loops here, one over fldlistFr and another over FldListTo?

File: src/drivers/nuopc/cime_flds/esmFldsExchange.F90 [hunk 14]
```diff
@@ -1109,62 +1112,79 @@ subroutine esmFldsExchange(gcomp, phase, rc)
...
+       call addfld(fldListFr(compatm)%flds, 'Faxa_snow' )
+       call addfld(fldListTo(compocn)%flds, 'Faxa_snow' )
+
+       call addfld(fldListFr(compatm)%flds, 'Faxa_snowc_wiso')
+       call addfld(fldListFr(compatm)%flds, 'Faxa_snowl_wiso')
+       call addfld(fldListFr(compatm)%flds, 'Faxa_snow_wiso' )
+       call addfld(fldListTo(compocn)%flds, 'Faxa_snow_wiso' )
     else
-       do n = 1,size(iso)
+       do n = 1,2
```
Comment (jedwards4b) [inline src/drivers/nuopc/cime_flds/esmFldsExchange.F90:1179]: seems like you are removing some flexibility and replacing it with hard coded values.

File: src/drivers/nuopc/mediator/med.F90 [hunk 15]
```diff
@@ -488,9 +489,9 @@ subroutine InitializeP0(gcomp, importState, exportState, clock, rc)
 
     call ESMF_LogWrite(trim(subname)//": Mediator verbosity is "//trim(value), ESMF_LOGMSG_INFO, rc=dbrc)
 
-!    dbug_flag = ESMF_UtilString2Int(value, &
-!         specialStringList=(/"min","max","high"/), specialValueList=(/0,255,255/), rc=rc)
-!    if (shr_nuopc_methods_ChkErr(rc,__LINE__,u_FILE_u)) return
+    ! dbug_flag = ESMF_UtilString2Int(value, &
```
Comment (jedwards4b) [inline src/drivers/nuopc/mediator/med.F90]: why not remove this?

**Review by jedwards4b — CHANGES_REQUESTED** [review]

File: src/drivers/nuopc/cime_driver/esmApp.F90 (hunk 9 above)
Comment (mvertens) [reply src/drivers/nuopc/cime_driver/esmApp.F90:150]: I have no idea - I have not changed this - should it not have a comment?

File: src/drivers/nuopc/cime_config/buildnml (hunk 7 above)
Comment (mvertens) [reply src/drivers/nuopc/cime_config/buildnml:341]: I agree totally. not sure when. First we need to get the performance up.

File: src/drivers/nuopc/cime_config/config_component_cesm.xml (hunk 8 above)
Comment (mvertens) [reply src/drivers/nuopc/cime_config/config_component_cesm.xml:206]: this will not effect the mct performance since this is in the nuopc/cime_config and NOT the mct/cime_config - they are totally separate

File: src/drivers/nuopc/mediator/med.F90 (hunk 15 above)
Comment (mvertens) [reply src/drivers/nuopc/mediator/med.F90]: good point - I will do that.

File: config/cesm/config_grids_common.xml (hunk 1 above)
Comment (mvertens) [reply config/cesm/config_grids_common.xml]: done

File: config/cesm/config_grids_common.xml (hunk 2 above)
Comment (mvertens) [reply config/cesm/config_grids_common.xml]: done

**Review by jedwards4b — APPROVED** [review]

**Comment by jedwards4b** [comment]
@uturuncoglu , @rsdunlapiv do either of you want to review before I merge?

**Review by uturuncoglu — APPROVED** [review]

**Review by rsdunlapiv — APPROVED** [review]
